wrappers/glmemshadow.cpp

```diff
@@ -0,0 +1,404 @@
+        if (it != ctx->dirtyShadows.end()) {
+            ctx->dirtyShadows.erase(it);
+        }
+    }
+
+    memProtect(shadowMemory, nPages * sPageSize, MemProtection::NO_ACCESS);
+
+    ctx = nullptr;
+    glMemory = nullptr;
+    flags = 0;
+    mappedStart = 0;
+    mappedSize = 0;
+    pagesToDirtyOnConsecutiveWrites = 1;
+}
+
+void GLMemoryShadow::onAddressWrite(uintptr_t addr, size_t page)
+{
+    const size_t relativePage = (addr - reinterpret_cast<uintptr_t>(shadowMemory)) / sPageSize;
+    if (isPageDirty(relativePage)) {
+        // It is possible if writing to the same buffer from two threads
+        return;
+    }
+
+    if ((relativePage == lastDirtiedRelativePage + 1) && isPageDirty(relativePage - 1)) {
+        /* Ensure that we would have log(n) page exceptions if traced application writes
+         * to n consecutive pages.
+         */
+        pagesToDirtyOnConsecutiveWrites *= 2;
+    } else {
+        pagesToDirtyOnConsecutiveWrites = 1;
+    }
+
+    const size_t endPageToDirty = std::min(relativePage + pagesToDirtyOnConsecutiveWrites, nPages);
+    for (size_t pageToDirty = relativePage; pageToDirty < endPageToDirty; pageToDirty++) {
+        setPageDirty(pageToDirty);
+    }
+
+    lastDirtiedRelativePage = endPageToDirty - 1;
+
+    memProtect(reinterpret_cast<void*>(page * sPageSize),
+               (endPageToDirty - relativePage) * sPageSize, MemProtection::READ_WRITE);
+}
+
+GLbitfield GLMemoryShadow::getMapFlags() const
+{
+    return flags;
+}
+
+void GLMemoryShadow::setPageDirty(size_t relativePage)
+{
+    assert(relativePage < nPages);
+    dirtyPages[relativePage / 32] |= 1U << (relativePage % 32);
+
+    if (!isDirty) {
+        ctx->dirtyShadows.push_back(this);
+        isDirty = true;
+    }
+}
+
+bool GLMemoryShadow::isPageDirty(size_t relativePage)
+{
+    assert(relativePage < nPages);
+    return dirtyPages[relativePage / 32] & (1U << (relativePage % 32));
+}
+
+void GLMemoryShadow::commitWrites(Callback callback)
+{
+    assert(isDirty);
+
+    uint8_t *shadowSlice = shadowMemory + mappedStartPage * sPageSize;
+    const size_t glStartOffset = mappedStart % sPageSize;
+
+    if (isPageDirty(mappedStartPage)) {
+        const size_t shadowOffset = mappedStart % sPageSize;
+        const size_t size = std::min(sPageSize - glStartOffset, mappedSize);
+
+        memcpy(glMemory, shadowSlice + shadowOffset, size);
+        callback(shadowSlice + shadowOffset, size);
+    }
+
+    for (size_t i = mappedStartPage + 1; i < mappedEndPage; i++) {
+        if (isPageDirty(i)) {
+            const size_t shadowOffset = (i - mappedStartPage) * sPageSize;
+            const size_t glOffset = shadowOffset - glStartOffset;
+            const size_t size = std::min(glStartOffset + mappedSize - shadowOffset, sPageSize);
+
+            memcpy(glMemory + glOffset, shadowSlice + shadowOffset, size);
+            callback(shadowSlice + shadowOffset, size);
```

Comment:
Rather that emitting one fake mempcy to the trace per page, it would be more efficient to coalesce all writes to consecutive pages into a single fake memcpy.